lib/LANraragi/Controller/Api/Archive.pm

```diff
@@ -107,6 +114,173 @@ sub serve_file {
+    my $expected_checksum   = $self->req->param('file_checksum'); # optional
+
+    # require file
+    if ( ! defined $upload || !$upload ) {
+        return $self->render(
+            json => {
+                operation   => "upload",
+                success     => 0,
+                error       => "No file attached"
+            },
+            status => 400
+        );
+    }
+
+    # checksum verification stage.
+    if ( $expected_checksum ) {
+        my $file_content        = $upload->slurp;
+        my $actual_checksum     = sha1_hex($file_content);
+        if ( $expected_checksum ne $actual_checksum ) {
+            return $self->render(
+                json => {
+                    operation   => "upload",
+                    success     => 0,
+                    error       => "Checksum mismatch: expected $expected_checksum, got $actual_checksum."
+                },
+                status => 417
+            );
+        }
+    }
+
+    my $filename        = $upload->filename;
+    my $uploadMime      = $upload->headers->content_type;
+
+    # utf downgrade (see LANraragi::Utils::Minion)
+    $filename = encode('UTF-8', $filename);
+    unless (utf8::downgrade($filename, 1)) {
+        $logger->error("Bullshit! File path \"$filename\" could not be converted back to a byte sequence!");
+        return $self->render(
+            json => {
+                operation   => "upload",
+                success     => 0,
+                error       => "\"$filename\" could not be converted back to a byte sequence!"
+            },
+            status => 422
+        )
+    };
+
+    # lock resource
+    my $lock            = $redis->setnx( "upload:$filename", 1 );
+    if ( !$lock ) {
+        return $self->render(
+            json => {
+                operation   => "upload",
+                success     => 0,
+                error       => "Locked resource: $filename."
+            },
+            status => 423
+        );
+    }
+    $redis->expire( "upload:$filename", 10 );
+
+    # metadata extraction
+    my $catid           = $self->req->param('category_id');
+    my $tags            = $self->req->param('tags');
+    my $title           = $self->req->param('title');
+    my $summary         = $self->req->param('summary');
+
+    # return error if archive is not supported.
+    if ( !is_archive($filename) ) {
+        $redis->del("upload:$filename");
+        $redis->quit();
+        return $self->render(
+            json => {
+                operation   => "upload",
+                success     => 0,
+                error       => "Unsupported file extension ($filename)"
+            },
+            status => 415
+        );
+    }
+
+    # Move file to a temp folder (not the default LRR one)
+    my $tempdir                 = tempdir();
+
+    my ( $fn, $path, $ext )     = fileparse( $filename, qr/\.[^.]*/ );
+    my $byte_limit              = LANraragi::Model::Config->enable_cryptofs ? 143 : 255;
```

Comment:
I wonder if this shouldn't be enforced at the `handle_incoming_file` level rather than both in here, in Model->Upload->Download_url, and the old Controller upload.  

Might be one for later tho

Reply:
Well, it's mostly there for synology hardware which has that 255 byte cap, and the notably insane who use cryptoFS on top of that, which limits it down to a further 143 bytes.  

I believe this would break even temporary filenames if exceeded.